src/main/java/com/useful/ucars/uCarsListener.java

```diff
@@ -1340,12 +1351,15 @@ public void run() {
 					}
 				}
 			}
-			if (fly && cont && (bidUpLoc.getBlock().getType().name().toLowerCase().contains("slab") || underblock.isEmpty() )) {
+			if (fly && cont && (bidUpLoc.getBlock().getType().name().toLowerCase().contains("slab") || underblock.isEmpty() && !carBlock.name().toLowerCase().contains("slab"))) {
 				// Make the car ascend (easter egg, slab elevator)
 				travel.setY(0.1); // Make a little easier
 				UEntityMeta.setMetadata(car, "car.ascending", new StatValue(null, plugin));
 			}
 			// Move the car and adjust vector to fit car stats
+			if(car.getLocation().getY() != car.getLocation().getBlockY()) {
+				travel.multiply(new Vector(0.65,1,0.65));
```

Comment:
Why are you reducing xz speed to 65% when y is not exactly equal to integer? (Which depending on the precision used to compare double (getY()) and int (getBlockY()) may never be true. I don't see why being on a block slightly elevated in y (eg. a slab) should arbitrarily slow down the car

Reply:
I do this to reduce/counter the effect of losing friction. When you were driving on slaps/climbing stuff/driving own something you were going quite a bit faster than when just driving on ground. Reducing the speed to 65% fixes this

Reply:
Hmm... I think this is an error and that this is just reducing speed of cars perhaps even everywhere for no good reason (The config should be used for that). If you are seeing a weird speed change it'll be due to a bug somewhere else (Perhaps multiplying xyz when meaning to only multiply xz components for example). I would wager this code is probably getting called even when driving on flat ground since 0.000000001 != 0

Reply:
If there is such a bug, seek to fix it rather than adding "bandaid" code to address only the symptoms

Reply:
I've done some testing and... I do not believe that this is an issue with ucars.

I let Minecraft move the Minecart with a vector of (1,0,0) - expectation: Just going into the X direction - Result: moving slightly downwards (seemingly y: -0.04) when moving through air.
I setup a "test-track" (straight line) and measured the time it took me to pass it. Slightly over 7 seconds.
I then tested (1,1,0) and (1,0.04,1) (the latter let's the minecart "glide" - no downward movement).
Both vectors resulted in a time of around 3.5 seconds on the track.

This to me looks like some sort of "friction", friction being pressure onto the ground, is being applied by Minecraft which is obviously missing when there is no "pressure" on the minecart.

Or am I misinterpreting something here.

Reply:
I've done some testing and... I do not believe that this is an issue with ucars.

I let Minecraft move the Minecart with a vector of (1,0,0) - expectation: Just going into the X direction - Result: moving slightly downwards (seemingly y: -0.04) when moving through air.
This behavior can also be observed when a Minecart is frozen (there was actually an issue with this which I found during testing but just setting the vector to 0 doesn't freeze you whilst (0,0.04,0) does)

I setup a "test-track" (straight line) and measured the time it took me to pass it. Slightly over 7 seconds.
I then tested (1,1,0) and (1,0.04,1) (the latter let's the minecart "glide" - no downward movement).
Both vectors resulted in a time of around 3.5 seconds on the track.

This to me looks like some sort of "friction", friction being pressure onto the ground, is being applied by Minecraft which is obviously missing when there is no "pressure" on the minecart.

Or am I misinterpreting something here.

(Tested on 1.16.5)

Reply:
Was this testing with ucars not installed?

Reply:
I tested it with and without uCars being installed, which is why I'm so confident that it's not an uCars error.

The adjustment only happens whilst climbing/going down slopes or whilst driving on non-whole blocks.
When driving on whole blocks car.getLocation().getY() and getBlockY() do actually return the same value.

One note:
I have another commit ready but I'm holding off of commiting it as long as this isn't resolved as that leaves me the option to change something if necessary without creating unnecessary commits

Reply:
Okay interesting. What about when driving on carpet? Or slabs? In those cases blockY is not the same as locationY? Should the "adjustment" (Reducing 'speed' requested from the game by 45%) be applied there or not? (I think it may still also be prudent to compare "Math.abs(blockY-locationY) < eps" instead of checking double and integer equality which is usually frowned upon as an additional tweak).

Reply:
Maybe also add a comment about why this is needed into the code (And move the "Move the car and adjust vector..." comment to be back next to the code it is relevant to)

Reply:
Will add the comments.

The speed-increase also happens whilst driving on carpet/slabs as we are kind of floating on them - the blockY-Location is always an integer.

But I have to admit that I don't know what you mean with eps - could you explain?

Reply:
Maybe a more robust check would be to check if we are asking for y velocity component > 0 or >= 0.04 in the velocity calculated by ucars and in that case we know the 'friction' is not applied. And the reverse aswell naturally.

And I know blockY location is always an integer, that's the problem. When you compare double and int the int will be promoted to a double for the comparison and it is generally not recommended to ever compare doubles for equality. By "eps" I meant "small threshold" (eps often refers to greek letter epsilon which is often used to denote small quantities in maths). The reason is that computers do not store floating point (even double precision floating point) numbers to infinite precision and 1.999999999 != 2.000000001
https://howtodoinjava.com/java-examples/correctly-compare-float-double/